src/dataflow/core/type_inference.py

```diff
@@ -0,0 +1,317 @@
+import json
+from abc import ABC
+from dataclasses import dataclass, field, replace
+from typing import Dict, List, Optional, Sequence, Set, cast
+
+from dataflow.core.definition import Definition
+from dataflow.core.program import (
+    BuildStructOp,
+    CallLikeOp,
+    Expression,
+    Program,
+    TypeName,
+    ValueOp,
+    roots_and_reentrancies,
+)
+
+
+@dataclass(frozen=True)
+class TypeInferenceError(Exception):
+    msg: str
+
+
+class Type(ABC):
+    pass
+
+
+class TypeVariable(Type):
+    pass
+
+
+class NamedTypeVariable(TypeVariable):
```

Comment:
Main changes all in here.

Reply:
Might be helpful to describe the algorithm at a high level somewhere, perhaps with references?

Reply:
Done